src/main/java/com/opentable/versionedconfig/GitOperations.java

```diff
@@ -53,30 +55,28 @@
 final class GitOperations {
     private static final Logger LOG = LoggerFactory.getLogger(GitOperations.class);
     private final Git git;
-    private final Optional<CredentialsProvider> credentials;
+    private final GitProperties config;
 
-    GitOperations(final GitProperties serviceConfig, Path checkoutDir) throws VersioningServiceException, IOException {
-        this.credentials = makeCredentials(serviceConfig.getUsername(), serviceConfig.getPassword());
-        this.git = openRepo(serviceConfig, checkoutDir);
+    GitOperations(final GitProperties config, Path checkoutDir) throws VersioningServiceException, IOException {
+        this.config = config;
+        this.git = openRepo(config, checkoutDir);
     }
 
-    private Optional<CredentialsProvider> makeCredentials(String username, String password) {
-        boolean missingUsername = isNullOrEmpty(username);
-        boolean missingPassword = isNullOrEmpty(password);
-
-        if (missingUsername && missingPassword) {
-            return Optional.empty();
-        } else if (missingUsername || missingPassword) {
-            throw new IllegalArgumentException("must provide username and password, or both must be empty/null");
+    void configureCredentials(TransportCommand<?, ?> op, URI uri) {
+        final String ui = uri.getUserInfo();
+        if (!StringUtils.isBlank(ui)) {
+            op.setCredentialsProvider(new UsernamePasswordCredentialsProvider(StringUtils.substringBefore(ui, ":"), StringUtils.substringAfter(ui, ":")));
         }
-        return Optional.of(new UsernamePasswordCredentialsProvider(username, password));
     }
 
     private Git openRepo(final GitProperties serviceConfig, Path checkoutDir)
             throws VersioningServiceException, IOException {
         if (Files.exists(checkoutDir) && Files.isDirectory(checkoutDir)) {
             LOG.info("checkout directory {} already exists", checkoutDir);
-            final boolean dirNotEmpty = Files.newDirectoryStream(checkoutDir).iterator().hasNext();
+            final boolean dirNotEmpty;
+            try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(checkoutDir)) {
+                dirNotEmpty = dirStream.iterator().hasNext();
+            }
             if (dirNotEmpty) {
                 return new Git(new FileRepository(checkoutDir.resolve(".git").toFile()));
```

Comment:
Log here too!

Reply:
This style of logging is pretty silly, it should just print once with what it did and why...